pkg/webhooks/controlplanemachineset/webhooks.go

```diff
@@ -60,17 +68,212 @@ func (r *ControlPlaneMachineSetWebhook) SetupWebhookWithManager(mgr ctrl.Manager
+)
+
 // ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
 func (r *ControlPlaneMachineSetWebhook) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	cpms, ok := obj.(*machinev1.ControlPlaneMachineSet)
+	if !ok {
+		return errTypeMismatchCondition
+	}
+
+	controlPlaneMachines, err := r.fetchControlPlaneMachines(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Ensure Control Plane Machine count matches the ControlPlaneMachineSet replicas
+	if int(*cpms.Spec.Replicas) != len(controlPlaneMachines) {
+		return fmt.Errorf("replicas: %v, Control plane machines: %v: %w", *cpms.Spec.Replicas, len(controlPlaneMachines), errControlPlaneReplicasDoesNotMatchMachineCount)
+	}
+
+	// Ensure CPMS created with invalid name is not allowed
+	if cpms.Name != "cluster" {
+		return fmt.Errorf("invalid name %s: %w", cpms.Name, errControlPlaneInvalidName)
+	}
+
+	// Ensure required labels are set and all machines are matching the label selector
+	if err := checkMachineLabels(cpms); err != nil {
+		return err
+	}
+
+	// Ensure failure domains of Control Plane Machines match the ControlPlaneMachineSet on create
+	if cpms.Spec.Template.OpenShiftMachineV1Beta1Machine.FailureDomains.Platform != "" {
+		if err := checkFailureDomains(cpms, controlPlaneMachines); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
 // ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
 func (r *ControlPlaneMachineSetWebhook) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) error {
+	oldCPMS, ok := oldObj.(*machinev1.ControlPlaneMachineSet)
+	if !ok {
+		return errTypeMismatchCondition
+	}
+
+	newCPMS, ok := newObj.(*machinev1.ControlPlaneMachineSet)
+	if !ok {
+		return errTypeMismatchCondition
+	}
+
+	// Ensure spec.Replicas is immutable on update
+	if *oldCPMS.Spec.Replicas != *newCPMS.Spec.Replicas {
+		return errControlPlaneReplicasChange
+	}
+
+	// Ensure selector is immutable on update
+	if !reflect.DeepEqual(oldCPMS.Spec.Selector, newCPMS.Spec.Selector) {
+		return errControlPlaneLabelSelectorChange
+	}
+
+	// Ensure required labels are set and all machines are matching the label selector
+	if err := checkMachineLabels(newCPMS); err != nil {
+		return err
+	}
+
 	return nil
 }
 
 // ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
 func (r *ControlPlaneMachineSetWebhook) ValidateDelete(ctx context.Context, obj runtime.Object) error {
 	return nil
 }
+
+// fetchControlPlaneMachines returns all control plane machines in the cluster.
+func (r *ControlPlaneMachineSetWebhook) fetchControlPlaneMachines(ctx context.Context) ([]*machinev1beta1.Machine, error) {
+	machineList := &machinev1beta1.MachineList{}
+	if err := r.c.List(ctx, machineList); err != nil {
+		return nil, fmt.Errorf("error querying api for machineList object: %w", err)
+	}
+
+	controlPlaneMachines := make([]*machinev1beta1.Machine, 0)
```

Comment:
You don't need to use make here, this will suffice
```suggestion
	controlPlaneMachines := []*machinev1beta1.Machine{}
```